Xamarin.Forms.Xaml/XamlParser.cs

```diff
@@ -329,32 +333,119 @@ static void GatherXmlnsDefinitionAttributes()
 		public static Type GetElementType(XmlType xmlType, IXmlLineInfo xmlInfo, Assembly currentAssembly,
 			out XamlParseException exception)
 		{
+			bool hasRetriedNsSearch = false;
+
+		retry:
 			if (s_xmlnsDefinitions == null)
 				GatherXmlnsDefinitionAttributes();
 
+			Type type = xmlType.GetTypeReference<Type>(
+				s_xmlnsDefinitions, 
+				currentAssembly?.FullName,
+				(asmName, clrNs, typeName) =>
+				{
+					return Type.GetType($"{clrNs}.{typeName}, {asmName}");
+				});
+			
+			var typeArguments = xmlType.TypeArguments;
+			exception = null;
+
+			if (type != null && typeArguments != null)
+			{
+				XamlParseException innerexception = null;
+				var args = typeArguments.Select(delegate(XmlType xmltype)
+				{
+					XamlParseException xpe;
+					var t = GetElementType(xmltype, xmlInfo, currentAssembly, out xpe);
+					if (xpe != null)
+					{
+						innerexception = xpe;
+						return null;
+					}
+					return t;
+				}).ToArray();
+				if (innerexception != null)
+				{
+					exception = innerexception;
+					return null;
+				}
+				type = type.MakeGenericType(args);
+			}
+
+			if (type == null) {
+#if NETSTANDARD2_0
+				// This covers the scenario where the AppDomain's loaded
+				// assemblies might have changed since this method was first
+				// called. This occurred during unit test runs and could
+				// conceivably occur in the field. 
+				if ( !hasRetriedNsSearch) {
+					hasRetriedNsSearch = true;
+					s_xmlnsDefinitions = null;
+					goto retry;
+				}
+#endif
+				exception = new XamlParseException(xmlType.GetErrorText(), xmlInfo);
+			}
+
+			return type;
+		}
+
+		public static string GetErrorText(this XmlType xmlType)
+		{
+			return string.Format(
+				"Type {0} not found in xmlns {1}. Ensure third party control " +
+				"libraries are referenced in the code of your project and not " +
+				"just in XAML.",
+				xmlType.Name,
+				xmlType.NamespaceUri);
+		}
+
+		public delegate T TypeFromXmlReference<T>(string assemblyName, string clrNamespace, string typeName)
+			where T : class;
+
+		public static T GetTypeReference<T>(
```

Comment:
FYI, this extension method is new as of yesterday because there would have otherwise been three separate places where this was being done almost identically . Same thing with the error message above.